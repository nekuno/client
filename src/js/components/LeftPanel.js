import React, { PropTypes, Component } from 'react';
import selectn from 'selectn';
import User from '../components/User';
import AuthenticatedComponent from '../components/AuthenticatedComponent';
import translate from '../i18n/Translate';
import connectToStores from '../utils/connectToStores';
import StatsStore from '../stores/StatsStore';
import LoginActionCreators from '../actions/LoginActionCreators';
import ChatThreadStore from '../stores/ChatThreadStore';
import ProfileStore from '../stores/ProfileStore';

function getState(props) {

    const stats = StatsStore.stats;
    const interests = selectn('numberOfContentLikes', stats) || 0;
    const unreadCount = ChatThreadStore.getUnreadCount() || 0;
    const profile = ProfileStore.get(props.user.id);

    return {
        interests,
        unreadCount,
        profile
    };
}

@AuthenticatedComponent
@translate('LeftPanel')
@connectToStores([StatsStore, ChatThreadStore], getState)
export default class LeftPanel extends Component {

    static contextTypes = {
        router: PropTypes.object.isRequired
    };

    static propTypes = {
        // Injected by @AuthenticatedComponent
        user        : PropTypes.object,
        userLoggedIn: PropTypes.bool,
        // Injected by @translate:
        strings     : PropTypes.object,
        // Injected by @connectToStores:
        interests   : PropTypes.number,
        unreadCount : PropTypes.number
    };

    constructor(props) {
        super(props);

        this.handleGoClickThreads = this.handleGoClickThreads.bind(this);
        this.handleGoClickProfile = this.handleGoClickProfile.bind(this);
        this.handleGoClickConversations = this.handleGoClickConversations.bind(this);
        this.handleGoClickSocialNetworks = this.handleGoClickSocialNetworks.bind(this);
        this.handleGoClickInterests = this.handleGoClickInterests.bind(this);
        this.handleClickMore = this.handleClickMore.bind(this);
        this.handleClickSettings = this.handleClickSettings.bind(this);
        this.handleGoClickInvitations = this.handleGoClickInvitations.bind(this);
        this.handleGoClickGroups = this.handleGoClickGroups.bind(this);
        this.logout = this.logout.bind(this);

        this.state = {
            moreActive: null
        };
    }

    handleGoClickThreads() {
        nekunoApp.closePanel();
        $$('.panel-left').once('closed', () => {
            this.context.router.push('/discover');
        });
    }

    handleGoClickProfile() {
        nekunoApp.closePanel();
        $$('.panel-left').once('closed', () => {
            this.context.router.push(`/p/${this.props.user.slug}`);
        });
    }

    handleGoClickConversations() {
        nekunoApp.closePanel();
        $$('.panel-left').once('closed', () => {
            this.context.router.push('/conversations');
        });
    }

    handleGoClickSocialNetworks() {
        nekunoApp.closePanel();
        $$('.panel-left').once('closed', () => {
            this.context.router.push('/social-networks');
        });
    }

    handleGoClickInterests() {
        nekunoApp.closePanel();
        $$('.panel-left').once('closed', () => {
            this.context.router.push('/interests');
        });
    }

    handleClickMore() {
        this.setState({
            moreActive: !this.state.moreActive
        });
    }

    handleClickSettings() {
        nekunoApp.closePanel();
        $$('.panel-left').once('closed', () => {
            this.context.router.push('/settings');
        });
    }

    handleGoClickInvitations() {
        nekunoApp.closePanel();
        $$('.panel-left').once('closed', () => {
            this.context.router.push('/invitations');
        });
    }

    handleGoClickGroups() {
        nekunoApp.closePanel();
        $$('.panel-left').once('closed', () => {
            this.context.router.push('/badges');
        });
    }

    logout() {
        nekunoApp.closePanel();
        $$('.panel-left').once('closed', () => {
            this.setState({
                moreActive: null
            });
            LoginActionCreators.logoutUser();
        });
    }

    render() {
        const {userLoggedIn, strings, interests, unreadCount} = this.props;
        const {moreActive} = this.state;
        return (
            <div className="LeftPanel">
                <div className="panel-overlay"></div>
                <div className="panel panel-left panel-reveal">
                    <div className="content-block top-menu">
                        <a className="close-panel">
                            <span className="icon-left-arrow"/>
                        </a>
                    </div>
                    { userLoggedIn ? <User {...this.props} onClick={this.handleGoClickProfile}/> : '' }
                    { userLoggedIn && !moreActive ?
                        <div className="user-interests">
                            <a href="javascript:void(0)" onClick={this.handleGoClickInterests}>
                                <div className="number">
                                    {interests}
                                </div>
                                <div className="label">
                                    {strings.interests}
                                </div>
                            </a>
                        </div>
                        : null
                    }
                    { userLoggedIn && !moreActive ?
                        <div className="content-block menu">
                            <a href="javascript:void(0)" onClick={this.handleGoClickThreads}>
                                <span className="icon-search"></span>&nbsp;&nbsp;{strings.threads}
                            </a>
                            {/*<a href="javascript:void(0)" onClick={this.handleGoClickProfile}>
                             {strings.myProfile}
                             </a>*/}
                            <a href="javascript:void(0)" onClick={this.handleGoClickConversations}>
                                <span className="icon-commenting"></span>&nbsp;&nbsp;{strings.conversations}
                                {unreadCount ? <span className="unread-messages-count">
                                    <span className="unread-messages-count-text">{unreadCount}</span>
                                </span> : ''}
                            </a>
                            <a href="javascript:void(0)" onClick={this.handleGoClickGroups}>
                                <span className="icon-puzzle-piece"></span>&nbsp;&nbsp;{strings.groups}
                            </a>
                            <a href="javascript:void(0)" onClick={this.handleClickMore}>
                                <span className="icon-fa-plus"></span>&nbsp;&nbsp;{strings.more}
                            </a>
                        </div>
                        : moreActive ?
                            <div className="content-block menu">
                                <a href="javascript:void(0)" onClick={this.handleClickMore} style={{fontWeight: 'bold'}}>
                                    <span className="icon-left-arrow"></span>&nbsp;&nbsp;{strings.less}
                                </a>
                                <a href="javascript:void(0)" onClick={this.handleGoClickProfile}>
                                    <span className="icon-person"></span>&nbsp;&nbsp;{strings.myProfile}
                                </a>
                                <a href="javascript:void(0)" onClick={this.handleGoClickSocialNetworks}>
                                    <span className="icon-plug"></span>&nbsp;&nbsp;{strings.socialNetworks}
                                </a>
                                <a href="javascript:void(0)" onClick={this.handleClickSettings}>
                                    <span className="icon-preferences"></span>&nbsp;&nbsp;{strings.settings}
                                </a>
                                {/*<Link to="/invitations" onClick={this.handleGoClickInvitations} onlyActiveOnIndex={false}>*/}
                                {/*{strings.invitations}*/}
                                {/*</Link>*/}
                                <a href="javascript:void(0)" onClick={this.logout}>
                                    <span className="icon-sign-out"></span>&nbsp;&nbsp;{strings.logout}
                                </a>
                            </div>
                            : '' }
                </div>
            </div>
        );
    }

}

LeftPanel.defaultProps = {
    strings: {
        interests     : 'Interests',
        threads       : 'Discover',
        groups        : 'Badges',
        myProfile     : 'Profile',
        conversations : 'Messages',
        socialNetworks: 'My social networks',
        more          : 'More',
        less          : 'Less',
        settings      : 'Settings',
        invitations   : 'Invitations',
        logout        : 'Logout'
    }
};