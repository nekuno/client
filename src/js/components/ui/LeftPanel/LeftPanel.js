import PropTypes from 'prop-types';
import React, { Component } from 'react';
import styles from './LeftPanel.scss';

export default class LeftPanel extends Component {

    static propTypes = {
        isOpen          : PropTypes.bool,
        handleClickClose: PropTypes.func.isRequired
    };

    handleClickClose() {
        this.props.handleClickClose();
    }

    render() {
        const {isOpen, children} = this.props;

        return (
            <div className={styles.leftPanelWrapper}>
                <div className={isOpen ? styles.open + ' ' + styles.leftPanel : styles.leftPanel}>
                    <div className={styles.contentWrapper}>
                        <div className={styles.content}>
                            {children}
                        </div>
                    </div>
                </div>
                <div className={isOpen ? styles.open + ' ' + styles.outsideWrapper : styles.outsideWrapper} onClick={this.handleClickClose.bind(this)}></div>
            </div>
        );
    }
}