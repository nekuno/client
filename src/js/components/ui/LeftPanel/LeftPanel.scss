@import '../../../../scss/_partials/colors';

.left-panel-wrapper {
  font-size: 15px;
  height: 100%;
  position: relative;
  width: 100%;
  z-index: 1111;

  .left-panel-absolute-wrapper {
    display: block;
    height:100vh;
    position: absolute;
    top: 0;
    width: 100%;
    overflow: hidden;

    .left-panel {
      background: #fff;
      display: inline-block;
      height: 100%;
      margin-left: -100%;
      width: 80%;
      float: left;
      position: relative;

      &.open {
        margin-left: 0;
        animation: open 0.25s linear;
      }

      .content {
        display: block;
        height: 100%;
        position: relative;
      }
    }

    .outside-wrapper {
      background: $regular-gray;
      display: inline-block;
      height: 100%;
      margin-left: -100%;
      opacity: 0.5;
      position: relative;
      width: 20%;

      &.open {
        margin-left: 0;
        animation: open-outside-wrapper 0.25s linear;
      }
    }
  }
}

@keyframes open {
  0% {
    transform: translate(-100%);
  }
  100% {
    transform: translate(0);
  }
}
@-webkit-keyframes open {
  0% {
    -webkit-transform: translate(-100%);
  }
  100% {
    -webkit-transform: translate(0);
  }
}
@keyframes open-outside-wrapper {
  0% {
    transform: translate(-500%);
  }
  100% {
    transform: translate(0);
  }
}
@-webkit-keyframes open-outside-wrapper {
  0% {
    -webkit-transform: translate(-500%);
  }
  100% {
    -webkit-transform: translate(0);
  }
}
