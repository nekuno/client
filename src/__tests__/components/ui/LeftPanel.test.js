import React from 'react';
import LeftPanel from '../../../js/components/ui/LeftPanel/LeftPanel.js';
import styles from '../../../js/components/ui/LeftPanel/LeftPanel.scss';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('Test LeftPanel component', () => {
    const mockFn = jest.fn();
    const closeLeftPanel = shallow(<LeftPanel handleClickClose={mockFn} isOpen={false}>Lorem ipsum</LeftPanel>);
    const openLeftPanel = shallow(<LeftPanel handleClickClose={mockFn} isOpen={true}>Lorem ipsum</LeftPanel>);

    it('should be defined', () => {
        expect(LeftPanel).toBeDefined();
    });
    it('should render correctly', () => {
        expect(closeLeftPanel).toMatchSnapshot();
        expect(openLeftPanel).toMatchSnapshot();
    });
    it('LeftPanel click event', () => {
        openLeftPanel.find(`.${styles.outsideWrapper}`).simulate('click');
        expect(mockFn).toHaveBeenCalledTimes(1);
    });
});
