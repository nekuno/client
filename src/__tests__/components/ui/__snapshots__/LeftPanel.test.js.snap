// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LeftPanel component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LeftPanel
    handleClickClose={[MockFunction]}
    isOpen={false}
  >
    Lorem ipsum
  </LeftPanel>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="leftPanelAbsoluteWrapper"
      >
        <div
          className="leftPanel"
        >
          <div
            className="content"
          >
            Lorem ipsum
          </div>
        </div>
        <div
          className="outsideWrapper"
          onClick={[Function]}
        />
      </div>,
      "className": "leftPanelWrapper",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="leftPanel"
          >
            <div
              className="content"
            >
              Lorem ipsum
            </div>
          </div>,
          <div
            className="outsideWrapper"
            onClick={[Function]}
          />,
        ],
        "className": "leftPanelAbsoluteWrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="content"
            >
              Lorem ipsum
            </div>,
            "className": "leftPanel",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Lorem ipsum",
              "className": "content",
            },
            "ref": null,
            "rendered": "Lorem ipsum",
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "outsideWrapper",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="leftPanelAbsoluteWrapper"
        >
          <div
            className="leftPanel"
          >
            <div
              className="content"
            >
              Lorem ipsum
            </div>
          </div>
          <div
            className="outsideWrapper"
            onClick={[Function]}
          />
        </div>,
        "className": "leftPanelWrapper",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="leftPanel"
            >
              <div
                className="content"
              >
                Lorem ipsum
              </div>
            </div>,
            <div
              className="outsideWrapper"
              onClick={[Function]}
            />,
          ],
          "className": "leftPanelAbsoluteWrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="content"
              >
                Lorem ipsum
              </div>,
              "className": "leftPanel",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Lorem ipsum",
                "className": "content",
              },
              "ref": null,
              "rendered": "Lorem ipsum",
              "type": "div",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "outsideWrapper",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Test LeftPanel component should render correctly 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LeftPanel
    handleClickClose={[MockFunction]}
    isOpen={true}
  >
    Lorem ipsum
  </LeftPanel>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="leftPanelAbsoluteWrapper"
      >
        <div
          className="open leftPanel"
        >
          <div
            className="content"
          >
            Lorem ipsum
          </div>
        </div>
        <div
          className="open outsideWrapper"
          onClick={[Function]}
        />
      </div>,
      "className": "leftPanelWrapper",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="open leftPanel"
          >
            <div
              className="content"
            >
              Lorem ipsum
            </div>
          </div>,
          <div
            className="open outsideWrapper"
            onClick={[Function]}
          />,
        ],
        "className": "leftPanelAbsoluteWrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="content"
            >
              Lorem ipsum
            </div>,
            "className": "open leftPanel",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Lorem ipsum",
              "className": "content",
            },
            "ref": null,
            "rendered": "Lorem ipsum",
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "open outsideWrapper",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="leftPanelAbsoluteWrapper"
        >
          <div
            className="open leftPanel"
          >
            <div
              className="content"
            >
              Lorem ipsum
            </div>
          </div>
          <div
            className="open outsideWrapper"
            onClick={[Function]}
          />
        </div>,
        "className": "leftPanelWrapper",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="open leftPanel"
            >
              <div
                className="content"
              >
                Lorem ipsum
              </div>
            </div>,
            <div
              className="open outsideWrapper"
              onClick={[Function]}
            />,
          ],
          "className": "leftPanelAbsoluteWrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="content"
              >
                Lorem ipsum
              </div>,
              "className": "open leftPanel",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Lorem ipsum",
                "className": "content",
              },
              "ref": null,
              "rendered": "Lorem ipsum",
              "type": "div",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "open outsideWrapper",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
